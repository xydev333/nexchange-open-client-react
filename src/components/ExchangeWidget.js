import React, { Component } from 'react';
import { Redirect, Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import axios from 'axios';
import _ from 'lodash';

import config from '../config';
import { fetchPrice, setWallet, errorAlert } from '../actions/index.js';

import CoinInput from './CoinInput';
import WalletAddress from './WalletAddress';


class ExchangeWidget extends Component {
	constructor(props) {
		super();
		
		this.state = {
			orderPlaced: false,
			loading: false,
	  	};
	  	  	 	
	  	this.placeOrder = this.placeOrder.bind(this);
	  	this.updatePrices = this.updatePrices.bind(this);
	}

	componentDidMount() {
		setTimeout(() => {
			this.updatePrices();
		}, 100);
	}

	componentWillUnmount() {
		clearTimeout(this.timeout);
	}

	updatePrices() {
		this.props.fetchPrice({pair: `${this.props.selectedCoin.receive}${this.props.selectedCoin.deposit}`, lastEdited: 'deposit', amount: this.props.amounts.deposit});

		this.timeout = setTimeout(() => {
			this.updatePrices();
		}, config.PRICE_FETCH_INTERVAL);
	}

	placeOrder() {
		this.setState({loading: true});

		axios({
			method: 'post',
			contentType : 'application/json',
			url: `${config.API_BASE_URL}/orders/`,
			data: {
				"amount_base": this.props.amounts.receive,
				"is_default_rule": true,
				"pair": {
					"name": `${this.props.selectedCoin.receive}${this.props.selectedCoin.deposit}`
				},
				"withdraw_address": {
					"address": this.props.wallet.address,
					"name": ""
				}
			}
		})
		.then(response => {
			this.setState({
				orderRef: response.data.unique_reference,
				orderPlaced: true,
				loading: false
			});
		})
		.catch(error => {
			console.log(error.response)

			let message = (error.response.data.non_field_errors && error.response.data.non_field_errors.length ? error.response.data.non_field_errors[0] : 'Something went wrong. Please try again later.');

			this.props.errorAlert({
				message: message,
				show: true,
				type: 'PLACE_ORDER'
			});

			this.setState({
				orderPlaced: false,
				loading: false,
			});
		});
	}

	componentWillReceiveProps(nextProps) {
		if (this.props.wallet.show && nextProps.error.type == 'INVALID_AMOUNT' && nextProps.error.show != false)
			this.props.setWallet({address: '', valid: false, show: false});
	}

	render() {
		if (this.state.orderPlaced)
			return <Redirect to={`/order/${this.state.orderRef}`} />

		return (
			<div className="col-xs-12">
				<div id="exchange-widget">
					<div className="col-xs-12 col-sm-6">
						<CoinInput type="deposit" />
					</div>

					<div className="col-xs-12 col-sm-6">
						<CoinInput type="receive" />
					</div>

					<WalletAddress />

					<div className="col-xs-12 text-center">
						{!this.props.wallet.show ? (
							<button className="btn btn-block proceed" onClick={() => this.props.setWallet({address: '', valid: false, show: true})} disabled={this.props.error.show && this.props.error.type == 'INVALID_AMOUNT' ? 'disabled' : null}>
								Get Started !
							</button>
						) : (
							<button className="btn btn-block proceed" onClick={this.placeOrder} disabled={(this.props.wallet.valid && !this.state.loading) ? null : 'disabled'}>
								Confirm & Place Order
								{this.state.loading ? <i className="fa fa-spinner fa-spin" style={{marginLeft: "10px"}}></i> : null}
							</button>
						)}
					</div>
				</div>
			</div>
		);
	}
}


function mapStateToProps(state) {
	return {
		selectedCoin: state.selectedCoin,
		amounts: state.amounts,
		error: state.error,
		wallet: state.wallet,
	}
}

function mapDispatchToProps(dispatch) {
	return bindActionCreators({
		fetchPrice: fetchPrice,
		setWallet: setWallet,
		errorAlert: errorAlert,
	}, dispatch)
}

export default connect(mapStateToProps, mapDispatchToProps)(ExchangeWidget);
