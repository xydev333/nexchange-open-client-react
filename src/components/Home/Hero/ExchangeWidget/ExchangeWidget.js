import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { I18n } from 'react-i18next';
import i18n from '../i18n';
import axios from 'axios';
import config from 'Config';

import { setWallet, errorAlert, setOrder } from 'Actions/index.js';
import { bindCrispEmail } from 'Utils/crispEmailBinding';

import CoinInput from './CoinInput/CoinInput';
import WalletAddress from './WalletAddress/WalletAddress';

import styles from './ExchangeWidget.scss';

class ExchangeWidget extends Component {
  constructor(props) {
    super();

    this.state = {
      orderPlaced: false,
      loading: false,
    };

    this.placeOrder = this.placeOrder.bind(this);
    this.showWalletAddress = this.showWalletAddress.bind(this);
  }

  componentWillUnmount() {
    clearTimeout(this.timeout);
  }

  placeOrder() {
    let data = {
      amount_base: 0,
      amount_quote: 0,
      is_default_rule: true,
      pair: {
        name: `${this.props.selectedCoin.receive}${this.props.selectedCoin.deposit}`,
      },
      withdraw_address: {
        address: this.props.wallet.address,
        name: '',
      },
    };

    if (this.props.price.lastEdited === 'receive') data['amount_base'] = parseFloat(this.props.price.receive);
    else if (this.props.price.lastEdited === 'deposit') data['amount_quote'] = parseFloat(this.props.price.deposit);

    axios
      .post(`${config.API_BASE_URL}/orders/`, data)
      .then(response => {
        this.props.setOrder(response.data);
        this.setState({
          orderRef: response.data.unique_reference,
          orderPlaced: true,
          loading: false,
        });

        if (response.data.token) {
          localStorage.setItem('token', response.data.token);
        }

        bindCrispEmail(this.props.store);

        window.ga('send', 'event', 'Order', 'place order', response.data.unique_reference);
        window.qp('track', 'Generic');
      })
      .catch(error => {
        console.log('Error:', error);

        let message =
          error.response && error.response.data.non_field_errors && error.response.data.non_field_errors.length
            ? error.response.data.non_field_errors[0]
            : `${i18n.t('subscription.5')}`;
        this.props.errorAlert({
          message: message,
          show: true,
          type: 'PLACE_ORDER',
        });

        this.setState({ orderPlaced: false, loading: false });
      });
  }

  showWalletAddress() {
    this.props.setWallet({ address: '', valid: false, show: true });

    setTimeout(() => {
      this.walletInputEl.focus();
    }, 300);
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    if (this.props.wallet.show && nextProps.error.type === 'INVALID_AMOUNT' && nextProps.error.show !== false) {
      this.props.setWallet({ address: '', valid: false, show: false });
    }
  }

  render() {
    if (this.state.orderPlaced) return <Redirect to={`/order/${this.state.orderRef}`} />;
    return (
      <div className="col-xs-12">
        <div className={styles.container}>
          <CoinInput type="deposit" onSubmit={this.showWalletAddress} />
          <CoinInput type="receive" onSubmit={this.showWalletAddress} />

          <WalletAddress onSubmit={this.placeOrder} inputRef={el => (this.walletInputEl = el)} />

        <I18n ns="translations">
         {(t) => (
          <div className="col-xs-12 text-center">
            {!this.props.wallet.show ? (
              <button
                className="btn btn-block btn-primary proceed"
                onClick={this.showWalletAddress}
                disabled={
                  this.props.error.show && (this.props.error.type === 'INVALID_AMOUNT' || this.props.error.type === 'INVALID_PAIR')
                    ? 'disabled'
                    : null
                }
              >
                {t('exchangewidget.1')}
              </button>
            ) : (
              <button
                className="btn btn-block btn-primary proceed"
                onClick={this.placeOrder}
                disabled={this.props.wallet.valid && !this.state.loading ? null : 'disabled'}
              >
                {t('exchangewidget.2')}
                {this.state.loading ? <i className="fab fa-spinner fa-spin" style={{ marginLeft: '10px' }} /> : null}
              </button>
            )}

            <p id="fee-info">The indicated price is final, all fees are included.</p>
          </div>
			)}
		  </I18n>
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ selectedCoin, price, error, wallet }) => ({
  selectedCoin,
  price,
  error,
  wallet,
});

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      setWallet: setWallet,
      setOrder: setOrder,
      errorAlert: errorAlert,
    },
    dispatch
  );

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ExchangeWidget);
