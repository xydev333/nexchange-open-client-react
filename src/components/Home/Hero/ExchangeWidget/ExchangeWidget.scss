.container {
  width: 100%;

  @media (min-height: $screen-lg-height) and (min-width: $screen-md-min) {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    padding-top: 100px;
  }
}

.widget {
  background: #fff;
  display: inline-block;
  width: 100%;
  color: #000;
  padding: 10px 0px;
  border-radius: 6px;
  text-align: left;

  @media (min-width: $screen-md-min) {
    padding: 36px 75px 35px;
  }
}

.submit {
  display: inline-block;
  width: 100%;
  padding: 0 15px;
  margin-top: 15px;
}

.info.info.info {
  color: $gray;
  font-size: 12px;

  @media (min-width: $screen-md-min) {
    margin-top: 13px;
    float: left;
  }
}

.btn.btn.btn {
  width: 100%;
  height: 48px;
  font-size: 12px;
  margin: 0;

  @media (min-width: $screen-md-min) {
    float: right;
    width: 238px;
  }
}
