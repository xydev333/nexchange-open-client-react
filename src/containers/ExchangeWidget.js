import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import axios from 'axios';
import config from '../config';

import { setWallet, errorAlert, setOrder } from '../actions/index.js';
import { bindCrispEmail } from '../helpers/crispEmailBinding';

import CoinInput from './CoinInput';
import WalletAddress from './WalletAddress';

class ExchangeWidget extends Component {
  constructor(props) {
    super();

    this.state = {
      orderPlaced: false,
      loading: false,
    };

    this.placeOrder = this.placeOrder.bind(this);
    this.showWalletAddress = this.showWalletAddress.bind(this);
  }

  componentDidMount() {
    $(function() {
      $('[data-toggle="tooltip"], [rel="tooltip"]').tooltip();
    });
  }

  componentWillUnmount() {
    clearTimeout(this.timeout);
  }

  placeOrder() {
    let data = {
      amount_base: 0,
      amount_quote: 0,
      is_default_rule: true,
      pair: {
        name: `${this.props.selectedCoin.receive}${this.props.selectedCoin.deposit}`,
      },
      withdraw_address: {
        address: this.props.wallet.address,
        name: '',
      },
    };

    if (this.props.price.lastEdited === 'receive') data['amount_base'] = parseFloat(this.props.price.receive);
    else if (this.props.price.lastEdited === 'deposit') data['amount_quote'] = parseFloat(this.props.price.deposit);

    axios
      .post(`${config.API_BASE_URL}/orders/`, data)
      .then(response => {
        this.props.setOrder(response.data);
        this.setState({
          orderRef: response.data.unique_reference,
          orderPlaced: true,
          loading: false,
        });

        if (response.data.token) {
          localStorage.setItem('token', response.data.token);
        }

        bindCrispEmail();

        window.ga('send', 'event', 'Order', 'place order', response.data.unique_reference);
        window.qp('track', 'Generic');
      })
      .catch(error => {
        let message =
          error.response && error.response.data.non_field_errors && error.response.data.non_field_errors.length
            ? error.response.data.non_field_errors[0]
            : 'Something went wrong. Please try again later.';
        this.props.errorAlert({
          message: message,
          show: true,
          type: 'PLACE_ORDER',
        });
        this.setState({ orderPlaced: false, loading: false });
      });
  }

  showWalletAddress() {
    this.props.setWallet({ address: '', valid: false, show: true });

    setTimeout(() => {
      this.walletInputEl.focus();
    }, 300);
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    if ($('#exchange-widget [data-toggle="tooltip"]').attr('aria-describedby')) {
      let tooltipId = $('#exchange-widget [data-toggle="tooltip"]').attr('aria-describedby');
      $(`#${tooltipId} .tooltip-inner`).html(`The fee amounts to ${nextProps.amounts.deposit * 0.005} ${nextProps.selectedCoin.deposit}.`);
    }

    if (this.props.wallet.show && nextProps.error.type === 'INVALID_AMOUNT' && nextProps.error.show !== false) {
      this.props.setWallet({ address: '', valid: false, show: false });
    }
  }

  render() {
    if (this.state.orderPlaced) return <Redirect to={`/order/${this.state.orderRef}`} />;

    return (
      <div className="col-xs-12">
        <div id="exchange-widget">
          <CoinInput type="deposit" onSubmit={this.showWalletAddress} />
          <CoinInput type="receive" onSubmit={this.showWalletAddress} />

          <WalletAddress onSubmit={this.placeOrder} inputRef={el => (this.walletInputEl = el)} />

          <div className="col-xs-12 text-center">
            {!this.props.wallet.show ? (
              <button
                className="btn btn-block btn-themed proceed"
                onClick={this.showWalletAddress}
                disabled={
                  this.props.error.show && (this.props.error.type === 'INVALID_AMOUNT' || this.props.error.type === 'INVALID_PAIR')
                    ? 'disabled'
                    : null
                }
              >
                Get Started !
              </button>
            ) : (
              <button
                className="btn btn-block btn-themed proceed"
                onClick={this.placeOrder}
                disabled={this.props.wallet.valid && !this.state.loading ? null : 'disabled'}
              >
                Confirm & Place Order
                {this.state.loading ? <i className="fa fa-spinner fa-spin" style={{ marginLeft: '10px' }} /> : null}
              </button>
            )}
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ selectedCoin, price, error, wallet }) => ({
  selectedCoin,
  price,
  error,
  wallet,
});

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      setWallet: setWallet,
      setOrder: setOrder,
      errorAlert: errorAlert,
    },
    dispatch
  );
}

export default connect(mapStateToProps, mapDispatchToProps)(ExchangeWidget);
