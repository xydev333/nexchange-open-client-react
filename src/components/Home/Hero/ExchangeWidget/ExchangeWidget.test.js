import React from 'react';
import { Provider } from 'react-redux';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import ExchangeWidget from './ExchangeWidget.js';

describe('ExchangeWidget', () => {
  const initialState = {
    wallet: { address: '', valid: false, show: false },
    selectedCoin: { deposit: 'OMG', receive: 'ETH', prev: { deposit: 'OMG', receive: 'ETH' }, lastSelected: 'deposit' },
    price: { pair: 'ETHOMG', deposit: 101, receive: 1.76250023, lastEdited: 'deposit' },
    error: { show: false, type: 'INVALID_AMOUNT' },
  };
  const mockStore = configureStore();
  let store, wrap;

  beforeEach(() => {
    store = mockStore(initialState);
    wrap = shallow(<ExchangeWidget store={store} />).dive();
  });

  it('renders correctly', () => {
    expect(wrap).toMatchSnapshot();
  });

  it('button changes on wallet shown', () => {
    expect(wrap.find('button').text()).toEqual('Get Started !');
    wrap.setProps({ wallet: { address: '', valid: false, show: true } });
    expect(wrap.find('button').text()).toEqual('Confirm & Place Order');
  });

  it('contains CoinInput, WalletAddress', () => {
    expect(wrap.find('Connect(CoinInput)').exists()).toBe(true);
    expect(wrap.find('Connect(WalletAddress)').exists()).toBe(true);
  });
});
