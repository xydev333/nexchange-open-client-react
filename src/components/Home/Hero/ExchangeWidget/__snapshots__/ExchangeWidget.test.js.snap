// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeWidget renders correctly 1`] = `ShallowWrapper {}`;
